File: searching/binary_search.cc
#include "utils/test.h"

namespace algorithm {
namespace searching {

template <class T, std::size_t N>
int32_t BinarySearch(T (&array)[N], T key, int32_t lo, int32_t hi) {
  if (lo > hi) {
    return -1;
  }

  int32_t mi = (lo + hi) / 2;
  if (array[mi] == key) {
    return mi;
  } else if (key < array[mi]) {
    return BinarySearch(array, key, lo, mi - 1);
  }
  return BinarySearch(array, key, mi + 1, hi);
}

namespace test {

int32_t ints[11] = {2, 3, 5, 6, 8, 9, 10, 22, 26, 32, 40};
std::string strs[7] = {"c", "cpp", "go", "java", "python", "sql", "swift"};
float flts[8] = {0.7f, 1.2f, 3.2f, 4.4f, 5.2f, 5.9f, 6.8f, 9.5f};

void BinarySearch_TestHandlesMultipleTypesOfArrayInput() {
  EXPECT_EQ(BinarySearch(ints, 8, 0, std::size(ints) - 1), 4);
  EXPECT_EQ(BinarySearch(strs, (std::string) "cpp", 0, std::size(strs) - 1), 1);
  EXPECT_EQ(BinarySearch(flts, 9.5f, 0, std::size(flts) - 1), 7);
}

void BinarySearch_TestHandlesNonExistElementInput() {
  EXPECT_EQ(BinarySearch(ints, 1, 0, std::size(ints) - 1), -1);
  EXPECT_EQ(BinarySearch(strs, (std::string) "rust", 0, std::size(strs) - 1),
            -1);
  EXPECT_EQ(BinarySearch(flts, 4.5f, 0, std::size(flts) - 1), -1);
}

void RunTests() {
  TEST(BinarySearch_TestHandlesMultipleTypesOfArrayInput);
  TEST(BinarySearch_TestHandlesNonExistElementInput);
}

}  // namespace test

}  // namespace searching
}  // namespace algorithm

int main() {
  std::cout << "-----Running binary search tests-----" << std::endl;
  algorithm::searching::test::RunTests();
  return 0;
}

File: searching/binary_search.py
"""
Binary search
"""
from typing import Any
import unittest


def binary_search(sorted_array: list, key: Any, lo: int, hi: int) -> int:
    if lo > hi:
        return -1
    mi = (lo + hi) // 2

    if sorted_array[mi] == key:
        return mi
    elif key < sorted_array[mi]:
        return binary_search(sorted_array, key, lo, mi - 1)
    return binary_search(sorted_array, key, mi + 1, hi)


class TestBinarySearch(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.integers = [2, 3, 5, 6, 8, 9, 10, 22, 26, 32, 40]
        self.strings = ["assembly", "c", "cpp", "go", "python", "sql", "swift"]
        self.floats = [0.7, 1.2, 3.2, 4.4, 5.2, 5.9, 6.8, 9.5]

    def test_handles_multiple_array_type_input(self):
        self.assertEqual(binary_search(self.integers, 8, 0, len(self.integers) - 1), 4)
        self.assertEqual(binary_search(self.strings, "c", 0, len(self.strings) - 1), 1)
        self.assertEqual(binary_search(self.floats, 9.5, 0, len(self.floats) - 1), 7)

    def test_handles_non_exist_element_input(self):
        self.assertEqual(binary_search(self.integers, 1, 0, len(self.integers) - 1), -1)
        self.assertEqual(
            binary_search(self.strings, "rust", 0, len(self.strings) - 1), -1
        )
        self.assertEqual(binary_search(self.floats, 4.5, 0, len(self.floats) - 1), -1)


if __name__ == "__main__":
    unittest.main()
